akka-stream/src/main/scala/akka/stream/KillSwitch.scala

```diff
@@ -245,7 +271,7 @@ final class SharedKillSwitch private[stream] (val name: String) extends KillSwit
     override def toString: String = s"SharedKillSwitchFlow(switch: $name)"
 
     override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, SharedKillSwitch) = {
-      val logic = new KillSwitches.KillableGraphStageLogic(shutdownPromise.future, shape) with InHandler with OutHandler {
+      val logic = new KillSwitches.KillableGraphStageLogic(shutdownPromises.weaklyLinkedFuture, shape) with InHandler with OutHandler {
```

Comment:
I always try hard to find another solution before using any of the weak things, since they are known to have a negative impact on garbage collectors.

Wouldn't it be possible to use something more rich internally that is passed as parameter to KillableGraphStageLogic instead of the Future. Then the KillableGraphStageLogic could use that to register and **deregister** the callback.

Reply:
:+1:

Reply:
I was trying to keep the amount of changes small and synchronization to a minimum. Modifying `KillableGraphStageLogic` will require changing `UniqueKillSwitchStage` to use the same machinery as well. If you are OK with making more changes, I can rebase with a more invasive solution that avoids `WeakRefs`.

Reply:
Understood. I think it is worth a try and using same machinery is good. Thanks